src/providers/wms/qgswmscapabilities.cpp

```diff
@@ -669,6 +669,44 @@ void QgsWmsCapabilities::parseCapability( QDomElement const &e, QgsWmsCapability
     }
   }
 
+  // Layers in a WMS-C Capabilities TileSet do not have layer title or abstract
+  // still those could be present in a Layer list.
+  if ( !mTileLayersSupported.isEmpty() )
+  {
+    QHash<QString, QString> titles;
+    QHash<QString, QString> abstracts;
+
+    for ( int i = 0; i < mLayersSupported.size(); ++i )
+    {
+      if ( !mLayersSupported.at( i ).title.isEmpty() && !mLayersSupported.at( i ).name.isEmpty() )
+      {
+        titles.insert( mLayersSupported.at( i ).name, mLayersSupported[i].title );
+      }
+    }
+    for ( int i = 0; i < mTileLayersSupported.size(); ++i )
+    {
+      if ( mTileLayersSupported.at( i ).title.isEmpty() && titles.contains( mTileLayersSupported.at( i ).identifier ) )
+      {
+        mTileLayersSupported[i].title = titles.value( mTileLayersSupported.at( i ).identifier );
+      }
+    }
+
+    for ( int i = 0; i < mLayersSupported.size(); ++i )
```

Comment:
Do you think it could be possible to fill the `abstracts` hash in the first loop at the same time as `titles`? It'll avoid a redundant loop.

Reply:
You are right - both loops can be combined. I have combined both loops and now code is much shorter and more elegant.